bip-0346.md

```diff
@@ -0,0 +1,438 @@
+```
+  BIP: 346
+  Layer: Consensus (soft fork)
+  Title: OP_TXHASH and OP_CHECKTXHASHVERIFY
+  Author: Steven Roose <steven@roose.io>, Brandon Black <freedom@reardencode.com>
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0346
+  Status: Draft
+  Type: Standards Track
+  Created: 2023-09-03
+  License: BSD-3-Clause
+```
+
+# Abstract
+
+This BIP proposes two new opcodes, `OP_CHECKTXHASHVERIFY`, to be activated
+as a change to the semantics of `OP_NOP4` in legacy script, segwit and tapscript;
+and `OP_TXHASH`, to be activated as a change to the semantics of `OP_SUCCESS189`
+in tapscript only.
+
+These opcodes provide a generalized method for introspecting certain details of
+the spending transaction, which enables non-interactive enforcement of certain
+properties of the transaction spending a certain UTXO.
+
+The constructions specified in this BIP also open up the way for other
+potential updates; see Motivation section for more details.
+
+As mentioned in the Deployment section, alternatively, the
+`OP_CHECKTXHASHVERIFY` semantics could be used to upgrade an existing deployment
+of the `OP_CHECKTEMPLATEVERIFY` opcode. The opcode `OP_TXHASH` could then
+potentially be renamed to `OP_TEMPLATEHASH`.
+
+
+# Summary
+
+## OP_CHECKTXHASHVERIFY
+
+The first new opcode, `OP_CHECKTXHASHVERIFY`, redefines the `OP_NOP4` opcode
+(`0xb3`) as a soft fork upgrade. This opcode is active in both legacy script,
+segwitv0 p2wsh and tapscript contexts.
+
+It has the following semantics:
+
+* There is at least one element on the stack, fail otherwise.
+* The element on the stack is at least 32 bytes long, NOP otherwise.
+* The first 32 bytes are interpreted as the StackTxHash and the remaining
+  suffix bytes specify the TxFieldSelector.
+* The TxFieldSelector valid, fail otherwise.
```

Comment:
Still digesting this proposal so apologies if bad question, but it might make sense for this opcode to trivially succeed if the tx field selector is not well defined by this BIP? This gives us flexibility in the future with extending how tx's are hashe? 🤷‍♂️

Reply:
That's not possible. We want to push something to the stack and going from nothing to the stack to something on the stack is a breaking change. Unfortunately it's not possible to have softfork-able hooks into this opcode.